src/libcore/iter/mod.rs

```diff
@@ -520,6 +520,35 @@ impl<I> Iterator for Cycle<I> where I: Clone + Iterator {
 #[unstable(feature = "fused", issue = "35602")]
 impl<I> FusedIterator for Cycle<I> where I: Clone + Iterator {}
 
+/// An iterator that skips n elements every iteration.
+///
+/// This `struct` is created by the [`every_nth`] method on [`Iterator`]. See
+/// its documentation for more.
+///
+/// [`every_nth`]: trait.Iterator.html#method.every_nth
+/// [`Iterator`]: trait.Iterator.html
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+#[stable(feature = "rust1", since = "1.0.0")]
+#[derive(Clone, Debug)]
+pub struct EveryNth<I> {
+    iter: I,
+    step: usize,
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I> Iterator for EveryNth<I> where I: Iterator {
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let elt = self.iter.next();
+        if self.step > 0 {
+            self.iter.nth(self.step - 1);
```

Comment:
At least in the basic use case of stepping through ranges, the compiler should be able to inline the whole thing and get rid of every abstraction (the fuse iterator or the `first_take` flag, `nth` etc). For other use cases, it might be impossible to optimise away the branch, but it might also be very bad to consume elements in advance.